pkg/controllers/provisioning/provisioner.go

```diff
@@ -92,18 +88,30 @@ type Provisioner struct {
 	packer        *binpacking.Packer
 }
 
+// Add a pod to the provisioner and block until it's processed. The caller
+// is responsible for verifying that the pod was scheduled correctly. In the
+// future, this may be expanded to include concepts such as retriable errors.
+func (p *Provisioner) Add(ctx context.Context, pod *v1.Pod) {
+	select {
+	case p.pods <- pod: // Block until pod is enqueued
+		<-p.wait
+	case <-p.running.Done(): // Leave if closed
+	}
+}
+
 func (p *Provisioner) provision(ctx context.Context) (err error) {
-	// Wait for a batch of pods
-	pods := p.Batch(ctx)
-	// Communicate the result of the provisioning loop to each of the pods.
+	// Wait for a batch of pods, release when done
+	pods := p.batch(ctx)
 	defer func() {
 		for i := 0; i < len(pods); i++ {
-			select {
-			case p.results <- err: // Block until result is communicated
-			case <-p.done: // Leave if closed
-			}
+			p.wait <- struct{}{}
```

Comment:
Previously, we would filter pods before releasing the threads, which resulted in miscounting when pods were bound.